nirum_wsgi.py

```diff
@@ -67,76 +125,87 @@ def route(self, environ, start_response):
         :param start_response:
 
         """
-        urls = self.url_map.bind_to_environ(environ)
         request = Request(environ)
-        try:
-            endpoint, args = urls.match()
-        except HTTPException as e:
-            return self.error(e.code, request)(environ, start_response)
+        service_methods = self.service.__nirum_service_methods__
+        error_raised = None
+        for _, pattern, verb, method_name in self.rules:
+            path_info = environ['PATH_INFO']
+            if isinstance(path_info, bytes):
+                # FIXME Decode properly; URI is not unicode
+                path_info = path_info.decode()
+            match = pattern.match(path_info)
+            if match and environ['REQUEST_METHOD'] == verb.upper():
+                routed = True
+                service_method = method_name
+                if verb in ('GET', 'DELETE'):
+                    method_parameters = {
+                        k: v
+                        for k, v in service_methods[method_name].items()
+                        if not k.startswith('_')
+                    }
+                    # TODO Parsing query string
+                    payload = {p: match.group(p) for p in method_parameters}
+                else:
+                    try:
+                        payload = parse_json_payload(request)
+                    except NoJsonError as e:
+                        error_raised = self.error(
+                            400, request,
+                            message="Invalid JSON payload: '{!s}'.".format(e)
+                        )
+                break
         else:
-            procedure = getattr(self, endpoint)
-            response = procedure(request, args)
-            return response(environ, start_response)
-
-    def ping(self, request, args):
-        return Response(
-            '"Ok"',
-            200,
-            content_type='application/json'
-        )
-
-    def rpc(self, request, args):
-        """RPC
-
-        :param request:
-        :args ???:
-
-        """
-        if request.method != 'POST':
-            return self.error(405, request)
-        payload = request.get_data(as_text=True) or '{}'
-        request_method = request.args.get('method')
-        if not request_method:
-            return self.error(
+            routed = False
+            if request.method != 'POST':
+                error_raised = self.error(405, request)
+            service_method = request.args.get('method')
+            try:
+                payload = parse_json_payload(request)
+            except NoJsonError as e:
+                error_raised = self.error(
+                    400, request,
+                    message="Invalid JSON payload: '{!s}'.".format(e)
```

Comment:
Same here.

> If we want to show the input payload it'd better to put it into the another field (e.g., `payload`) rather than embed it in `message` text.